File: android/src/main/java/com/lightningkite/khrysalis/Video.actual.kt
package com.lightningkite.khrysalis

import android.graphics.PointF
import android.media.MediaMetadataRetriever
import android.os.Build
import com.lightningkite.khrysalis.net.HttpClient
import com.lightningkite.khrysalis.views.ViewStringRaw
import com.lightningkite.khrysalis.views.loadVideoThumbnail
import com.lightningkite.khrysalis.views.showDialog
import io.reactivex.Single
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers

fun Video.thumbnail(timeMs: Long = 2000L, size: PointF? = null): Single<Image> {
    return Single.create<Image> { em ->
        try {
            val mMMR = when (this) {
                is VideoReference -> {
                    val mMMR = MediaMetadataRetriever()
                    mMMR.setDataSource(HttpClient.appContext, this.uri)
                    mMMR
                }
                is VideoRemoteUrl -> {
                    val mMMR = MediaMetadataRetriever()
                    mMMR.setDataSource(this.url, HashMap<String, String>())
                    mMMR
                }
            }
            if (size != null && Build.VERSION.SDK_INT >= 27) {
                em.onSuccess(
                    ImageBitmap(
                        mMMR.getScaledFrameAtTime(
                            timeMs ,
                            MediaMetadataRetriever.OPTION_CLOSEST_SYNC,
                            size.x.toInt(),
                            size.y.toInt()
                        )
                    )
                )
            } else {
                em.onSuccess(ImageBitmap(mMMR.getFrameAtTime(timeMs)))
            }
        } catch (e: Exception) {
            em.onError(e)
        }
    }
        .subscribeOn(Schedulers.io())
        .observeOn(AndroidSchedulers.mainThread())
}

File: ios/Khrysalis/src/Video.actual.swift
import AVKit
import UIKit
import RxSwift

public extension Video {
    func thumbnail(timeMs: Int64 = 2000, size: CGPoint? = nil) -> Single<Image> {
        return Single.create { (em: SingleEmitter<Image>) in
            let vid: AVAsset
            switch self {
            case let self as VideoRemoteUrl:
                guard let url = URL(string: self.url) else {
                    em.onError(IllegalArgumentException("Invalid video URL \(self.url)", nil))
                    return
                }
                vid = AVAsset(url: url)
            case let self as VideoReference:
                vid = AVAsset(url: self.uri)
            default:
                em.onError(IllegalArgumentException("Invalid video type \(self)", nil))
                return
            }
            let imageGenerator = AVAssetImageGenerator(asset: vid)
            imageGenerator.appliesPreferredTrackTransform = true
            let time = CMTime(seconds: Double(timeMs) / 1000.0, preferredTimescale: 600)
            let times = [NSValue(time: time)]
            imageGenerator.generateCGImagesAsynchronously(forTimes: times, completionHandler: { _, image, _, _, _ in
                if let image = image {
                    em.onSuccess(ImageBitmap(bitmap: UIImage(cgImage: image)))
                } else {
                    em.onError(IllegalStateException("Could not generate preview", nil))
                }
            })
        }.subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread())
    }
}

